### PR TITLE
8277531: Print actual default stacksize on Windows thread logging

### PR DESCRIPTION
**Pull request by tkiriyama**
I modified to log actual stack size on Windows by using `os::win32::default_stack_size()`.
Could you please review this fix?

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277531](https://bugs.openjdk.java.net/browse/JDK-8277531): Print actual default stacksize on Windows thread logging

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 0bdc2fb69321e32e101ec7e80a9a711250769b76
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6495/head:pull/6495` \
`$ git checkout pull/6495`

Update a local copy of the PR: \
`$ git checkout pull/6495` \
`$ git pull https://git.openjdk.java.net/jdk pull/6495/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6495`

View PR using the GUI difftool: \
`$ git pr show -t 6495`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6495.diff">https://git.openjdk.java.net/jdk/pull/6495.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/logging/TestLogThreadStackSize.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8277531
+ * @requires os.family == "windows"
+ * @summary os+thread output should contain actual thread size for thread start
+ * @library /test/lib
+ */
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/logging/TestLogThreadStackSize.java:30]: Test should use:

`@run driver TestLogThreadStackSize`

**Review by dholmes-ora — COMMENTED** [review]
This change seems reasonable, but it appears to me that default_stack_size() actually reports the stack size of the initial thread, which AFAIK could be modified by passing -Xss, so it would not be the default afterall.

David

**Comment by dholmes-ora** [comment]
Though perhaps that is expected - what we actually intend to print is the -Xss value? But if so why not just print that?

Also with current logging and -Xss256K I see a strange mix of "default" and "256K" in the output, which seems wrong in relation to the current patch. I will need to look at this more closely.

David

**Comment by dholmes-ora** [comment]
@tkiriyama apologies for my confusion over this. So ... 

On Posix systems we read back the value of the stackSize from the pthread-attr object used to create the pthread. Very simple and accurate. 

On Windows there is no direct way to query the stackSize of the thread that was just created. Instead we used the requested stackSize: if it is > 0 then we assume we get what we asked for. If it is zero then it means use the VM default stackSize for JavaThreads - which is `ThreadStackSize` and which can be set directly or by `-Xss`. If that value is > 0 then again we assume the size of the stack we get is the size that we asked for. If `ThreadStackSize` itself is zero (the default value for Windows) that means "tell the operating system to use its own default" - and that is where we currently will print `stacksize: default` because we don't know what that operating system default is.

What this patch does is replace "default" with the value from `os::win32::default_stack_size()` which is set by calling `os::current_stack_size()` on the main thread during VM initialization. As the main thread is created by the launcher using the `-Xss` value (if set) the main thread's stack does represent the "default stack size" for all Java threads. So that sounds good.

The problem I have here is not with this patch per se, it is that `os::current_stack_size()` also has no direct way to query the stack size of the current thread. What it does is to use `VirtualQuery` passing in a stack allocated data structure, and asks for the `AllocationBase` for that structure. That in turn is the base address for the set of pages allocated by `VirtualAlloc` that are contiguous to and have the same state (regarding mapping and protection etc) as the address on the stack. So if the stack is allocated via a `VirtualAlloc` call as one coherent set of pages, then this gives us the base address of that set of pages. We then subtract that value from the one obtained from `os::current_stack_base()` to get the stack size. Where `os::current_stack_base()` itself uses `VirtualQuery` in a similar way to try to find the other end of this page range. The process is quite complex - more complex than I would expect as it tries to account for distinct regions that all have the same `AllocationBase` - which is not something that really makes sense to me. But at the end of that we will have a number that at least approximates the size of the stack of the main thread, and hence the default stack size.

It would be interesting (and I will try it out) to see what `os::win32::default_stack_size()` actually reports for different `-Xss` values.

Meanwhile I will approve the fix, but please update the test as requested. You will also need a second reviewer.

Thanks,
David

**Review by dholmes-ora — APPROVED** [review]
For the record on my Windows 10 system, the default stacksize (-Xss0) is 1024K and any requested -Xss value is rounded up to a 64K boundary.

Thanks,
David

**Comment by tstuefe** [comment]
Hi,

I try to understand this:

- in libjli we parse Xss the first time
- create main thread with stacksize=Xss
- invoke CreateJavaVM
- in `os::init_2` we do those VirtualQuery() games to sniff out the thread stack size of the main thread, store that as `win32::_default_stack_size`
- When creating new threads, we have the switch construct in `os::create_thread` (lines 695ff) where we set stack_size of new thread depending on thread type (which, btw, is what all posix platforms do in their implementation of `os::default_stack_size()`). But it can be left zero. If left zero, we call _beginthreadex with stack_size=0. 
- We then call `describe_beginthreadex_attributes()` with stack_size=0. With this patch it would then print out the value of `win32::_default_stack_size`. I don't get the logic. What would this correspond to whatever Xss was? We just told the system to create a stack with whatever it thinks the default stack size should be. Whereas in jli, we explicitly specified a stack size.

So, I am not sure yet this is correct. 

I would like to see tests where we explicitly specify `Xss`, `CompilerThreadStackSize` and `VMThreadStackSize` and make sure correct values show up in the log. Sorry, a wrong output on logging is worse than no output at all.

---

Zooming out, unrelated to this specific patch: I dislike how `os::default_stack_size()` mean different things in Posix and Windows. On Posix, it is a constant defined by us. On windows, it is whatever the main thread had been started with, plus some VirtualQuery related fuzziness. Currently, I am not sure what use this is on Windows.

On windows, `os::default_stack_size()` seems unused, at least I'm not sure what it is used for. The only use I find is to calculate `win32::_os_thread_limit` (see `os::init_2`), but that itself seems unused. What would be nice is to unify logic of `os::default_stack_size()` across all platforms.

I may be missing something here.

---

About `os::current_stack_size()`:

I am not 100% sure sure the VirtualQuery is always correct. When doing some work with VirtualQuery (see https://bugs.openjdk.java.net/browse/JDK-8255978) I saw that the kernel merges back-to-back segments, similar to what Linux does with its VMAs. So what we think the thread stack is may actually be a number of fused regions showing up as one. Maybe it is not an issue in reality though as side effect of ours or the systems stack guards.

**Comment by dholmes-ora** [comment]
@tstuefe I had hoped my comment made this clearer. But we do have to look at the main thread created by the launcher separate from threads created by the VM to ensure the process is the same. If there is no -Xss set the libjli code will pass zero as the stacksize to ContinueInNewThread (else the Xss value). If that code receives zero as the stacksize it checks if the vm_args have set an explicit `javaStackSize`, and if so that is used. Else zero is passed through as the stacksize for _beginthreadex and the stack size is determined by Windows.

The main thread then does the VirtualQuery dance that tries to determine what the main threads actual size is and that is set as the os:;win32::default_stack_size() and represents the stacksize for any thread for which zero is passed to beginthreadex.

I too have some queries over the use of VirtualQuery but note that os::current_stack_base does loop through checking for adjacent regions (not that I fully understand it).

But all of this is somewhat of an aside to this PR. We have a function that reports what the stack size is for threads which pass zero to beginThreadex, and this PR simply uses that function to report a number instead of the word "default" when the log the thread info. We can argue about that function's operation but that is a separate issue.

Cheers,
David

**Comment by tstuefe** [comment]
>@tstuefe I had hoped my comment made this clearer. But we do have to look at the main thread created by the launcher separate from threads created by the VM to ensure the process is the same. If there is no -Xss set the libjli code will pass zero as the stacksize to ContinueInNewThread (else the Xss value). If that code receives zero as the stacksize it checks if the vm_args have set an explicit javaStackSize, and if so that is used. Else zero is passed through as the stacksize for _beginthreadex and the stack size is determined by Windows.

>The main thread then does the VirtualQuery dance that tries to determine what the main threads actual size is and that is set as the os:;win32::default_stack_size() and represents the stacksize for any thread for which zero is passed to beginthreadex. (Actually it will also report the size of a non-zero -Xss - it reports whatever stack size the main thread got.)

The latter was my point. E.g. what about Xss != 0 and CompilerThreadStackSize == 0. Wouldn't we then start compiler threads with stack_size=0? Let the system decide? But the we'd go into the "default" reporting case and report `os::default_stack_size()` which is whatever the main thread uses after having been started non-0 Xss.

>But all of this is somewhat of an aside to this PR. We have a function that reports what the stack size is for threads which pass zero to beginThreadex

No, we have a function which reports the stack size of threads started with whatever Xss was, which may not be the system default, and may not match the stack size of threads started with stack size zero.

Sorry for being difficult. The patch is useful. But we should be able to rely on the numbers. That's why I would like more thorough tests which actually test the log output semantically, for a combination of stack size arguments, as proposed previously.

Thanks!

Thomas

**Comment by dholmes-ora** [comment]
@tstuefe I see your point. We should only use os::win32::default_stack_size() when printing JavaThreads that are not CompilerThreads. For CompilerThreads we have no way to know what the actual stacksize was if zero was specified.

**Comment by dholmes-ora** [comment]
There are also issues if you combine -Xss with -XX:ThreadStackSize as the -Xss will be applied to the main thread by the launcher, but the ThreadStackSize will be applied to subsequent JavaThreads. But there is no way to tell that at present. This combination is treated as multiple occurrences of ThreadStackSize with last-option wins - except the launcher will already have only looked at -Xss.

**Comment by tstuefe** [comment]
> There are also issues if you combine -Xss with -XX:ThreadStackSize as the -Xss will be applied to the main thread by the launcher, but the ThreadStackSize will be applied to subsequent JavaThreads. But there is no way to tell that at present. This combination is treated as multiple occurrences of ThreadStackSize with last-option wins - except the launcher will already have only looked at -Xss.

I wondered about that too. We should fix this in the launcher. A bit of a mess :-/

We could spawn off a stack-size=0 thread, just to measure its stack size. But that would cost startup and feels a bit ridiculous tbh.

@tkiriyama : none of this is your fault, the patch would be fine if `os::default_stack_size()` would do what it advertises...

**Comment by tkiriyama** [comment]
Thank you so much for your review.
Sorry it took me so long to understand discussion.

Is there something wrong to print out the value of default_stack_size?
I could not confirm the problem when I combine -Xss with -XX:ThreadStackSize., but it doesn't affect this patch.

I will try to fix this test for -Xss, -XX:ThreadStackSize, -XX:CompilerThreadStackSize and -XX:VMThreadStackSize.
Please let me know if there are any other options which need to be tested.

**Comment by tkiriyama** [comment]
> The problem is that os:;win32::default_stack_size() doesn't necessarily
> report the default stack size - it simply reports the stackSize of the
> initial main JavaThread thread. So if you print it, it should only be
> for JavaThreads that are not CompilerThreads.

In either case of JavaThreads or CompilerThreads,the stack_size argument of _beginthreadex() must be 0 when this code is executed.
According to MSDN document, If the stack_size argument can be 0, the operating system uses the same value as the stack that's specified for the main thread.
default_stack_size is specified in create_vm() which runs in the main thread, so we use default_stack_size() for CompilerThreads.

> But if you run "java -Xss256k -XX:ThreadStackSize=512K" it will print
> 256K as the default stack size when in fact all JavaThreads except the
> main thread will get a 512K stack because of ThreadStackSize.

Yes,this behavior is odd. However, this fix executed only if stacksize is not specified.
This problem should be fixed in another issue.

**Comment by tstuefe** [comment]
Hi, @tkiriyama,

> > The problem is that os:;win32::default_stack_size() doesn't necessarily
> > report the default stack size - it simply reports the stackSize of the
> > initial main JavaThread thread. So if you print it, it should only be
> > for JavaThreads that are not CompilerThreads.
> 
> In either case of JavaThreads or CompilerThreads,the stack_size argument of _beginthreadex() must be 0 when this code is executed. According to MSDN document, If the stack_size argument can be 0, the operating system uses the same value as the stack that's specified for the main thread. default_stack_size is specified in create_vm() which runs in the main thread, so we use default_stack_size() for CompilerThreads.

Nope, unfortunately not. It runs on the thread invoking CreateJavaVM(). That is usually not the OS main thread (aka primordial thread). In the case of the standard java launcher, the primordial thread spawns a second thread to invoke CreateJavaVM() in, mainly to impose -Xss. In the case of a custom launcher, all bets are open which thread invokes CreateJavaVM(), but typically it's not the primordial thread either. In fact, we have platforms that explicitly forbid invoking CreateJavaVM() on the primordial thread.

**Comment by tkiriyama** [comment]
Hi, @tstuefe ,

It’s exactly as you said.  
I tried to change _default_stack_size to be set in the main thread or get the stack size of other thread, but it doesn't work well.
I wonder if there might be another way to do this.

**Comment by tstuefe** [comment]
> Hi, @tstuefe ,
> 
> It’s exactly as you said. I tried to change _default_stack_size to be set in the main thread or get the stack size of other thread, but it doesn't work well. I wonder if there might be another way to do this.

I only have bad ideas:

Bad idea:
1) start a thread on windows with stacksize 0. Measure stack size. Now you know the stack size. Obviously bad since we pay for the startup of one whole thread, pointlessly. Whose only point of existing is to tell us its stack size. 

Slightly better idea, but still bad:
2) set _default_stack_size=0 meaning unknown. The first thread created with stacksize=0 measures its own stack and updates that variable. Obviously bad because: too much complexity (e.g. how does the child know it was created with stacksize=0), and at least the first stacksize=0 thread will still be printed with "stacksize=0" by its parent.

Also bad, but differently bad :-) :
3) Just measure default stack size on a developer machine, assume it is the same as on every runtime machine. Hard-code that into the VM. Obviously also bad.

So, for now I don't see a good idea. If you come up with something better, feel free. But its not a big problem tbh, so probably not worth many brain cycles.

Cheers, Thomas

**Comment by tkiriyama** [comment]
Hi, @tstuefe , 

Thank you for your ideas.  
I thought about some alternatives for fix, but unfortunately it is diffcult to print actual stack size when the thread is created.

It is possible to get the stack size when the thread is alive.
I can print actual stack size by thread->stack_size() in thread_native_entry().

With this fix, additional information is printed when logging "thread is alive".
Log format of Xlog changes, but it may be useful that we can know actual stack size in this log.

**Comment by dholmes-ora** [comment]
I think adding the stacksize to the "thread is alive" log statement is a reasonable alternative approach. The JBS issue should be updated to reflect the change in approach. I don't think we need a new test for this but you could possibly add something to the existing ./runtime/logging/ThreadLoggingTest.java

**Comment by tkiriyama** [comment]
I think so too.
I fixed runtime/logging/ThreadLoggingTest.java for this fix.
I can’t comment to JBS, so I ask someone.

**Review by dholmes-ora — APPROVED** [review]
I'm okay with seeing this go ahead as-is. Please change the second copyright year from 2021 to 2022 in the source file and test file.

Arguably all platforms could include the stack size in this part of the logging, but I'm okay with just Windows for now. (I think there is room for sharing a lot of the thread startup logic on non-Windows, at which time I'd add the logging change.)

Thanks,
David

**Comment by tkiriyama** [comment]
Thank you for your review.
I changed the copyright year from 2021 to 2022 in the source file and test file.

**Comment by tkiriyama** [comment]
@dholmes-ora 
I hope this change is integrated.  Can I ask you to sponsor?

**Comment by dholmes-ora** [comment]
Two reviews are needed for hotspot changes, so we need to wait to see if @tstuefe is okay with this.

**Review by tstuefe — APPROVED** [review]
I think this looks fine. Sorry for losing track on this PR.

Cheers, Thomas

**Pull request closed without merging**